web-server-lib/web-server/http/request.rkt

```diff
@@ -274,6 +274,31 @@
 (module+ internal-test
   (provide read-http-line/limited))
 
+; read-bytes/lazy : number inp -> bytes
+; Like `read-bytes', but waits until the expected number of bytes is
+; available within the input port before allocating the final buffer.
+(define (read-bytes/lazy n in [bufsize 4096])
+  (define buf (make-bytes bufsize))
+  (define offset
+    (let loop ([offset 0])
+      (define len
+        (peek-bytes-avail! buf offset #f in))
+
+      (cond
+        [(eof-object? len) offset]
+        [else
+         (define offset* (+ offset len))
+         (cond
+           [(> offset* n) offset*]
+           [else (loop offset*)])])))
+
+  (cond
+    [(zero? offset) eof]
+    [else (read-bytes n in)]))
```

Comment:
@Bogdanp If I am following correctly (a big "if"), I think this implementation of `read-bytes/lazy` will busy-wait until enough bytes become available, whereas the old `peek-bytes-evt`-based version yielded to the Racket thread scheduler until it could make some progress. If I'm right about that, is it something you considered?

I wouldn't hold <https://github.com/racket/web-server/pull/84> for it in any case, since it isn't an API change, but on general principles I would think yielding to Racket would be better. If so, we should probably use <https://github.com/racket/racket/commit/c19b944> and add an appropriate `#:version` requirement to `info.rkt`.

Reply:
@LiberalArtist `peek-bytes-avail` blocks when there isn't enough data available on the port so the code does cooperate with the rest of Racket fine. Just to be sure I ran this test on my machine:

```racket
#lang racket/base

(define-values (in out) (make-pipe))

(define (read-bytes/lazy n in [bufsize 4096])
  (define buf (make-bytes bufsize))
  (define offset
    (let loop ([offset 0])
      (displayln "reading")
      (define len
        (peek-bytes-avail! buf offset #f in))

      (printf "read ~a bytes\n" len)
      (cond
        [(eof-object? len) offset]
        [else
         (define offset* (+ offset len))
         (cond
           [(> offset* n) offset*]
           [else (loop offset*)])])))

  (cond
    [(zero? offset) eof]
    [else (read-bytes n in)]))

(thread
 (lambda _
   (read-bytes/lazy 1024 in)))

(for ([_ (in-range 1024)])
  (displayln "writing")
  (display "a" out)
  (sleep 1))
```

which yields:

```
writing
reading
read 1 bytes
reading
writing
read 1 bytes
reading
writing
...
```

I don't think the commit you linked to would help here. The reason I made this function is that all of the standard reading functions pre-allocate the result buffers based on the input size they're given. This means that calling `(read-bytes 1024 some-port)` immediately allocates a 1k buffer whether there is 1k of data on the port or not. This is problematic when dealing with user input because someone could send a `Content-Length: 10000000000000` and crash our server. Or, presuming we limit the `Content-Length` to some max, they could send many requests with `Content-Length: max-1` without ever sending any body data.

Reply:
I see; I was confused about `peek-bytes-avail!`. Thanks!